### DISCUSSION TITLE
Retries and decorating checked suppliers issue

### DISCUSSION THREAD

**Discussion by malenloth** [resilience4j/resilience4j#1977] [Q&A]
Hi,

We've been using resilience4j in our project for a while, and now we're facing some issues when trying to migrate our code to 2.0.2 version that we can't seem to figure out. 

https://resilience4j.readme.io/docs/retry#decorate-and-execute-a-functional-interface still has the old code as example (even with v2.0.0 selected), before the vavr dependency has been removed. The code we're having trouble with is essentially the same as this example:

> // Decorate the invocation of the HelloWorldService
> CheckedFunction0<String> retryableSupplier = Retry
>   .decorateCheckedSupplier(retry, helloWorldService::sayHelloWorld);
> 
> // When I invoke the function
> Try<String> result = Try.of(retryableSupplier)
>   .recover((throwable) -> "Hello world from recovery function");

Retry#decorateCheckedSupplier now returns CheckerSupplier, not CheckedFunction0, and CheckedSupplier does not extend Supplier, so it cannot be passed to Try#of or Try#ofSupplier

Documentation here in gitlab has the following example for retries, again using vavr's Try class, but in this case we wouldn't be able to throw a checked exception in the Supplier:

> Supplier<String> decoratedSupplier = CircuitBreaker
>     .decorateSupplier(circuitBreaker, backendService::doSomething);
> 
> // Decorate your call with automatic retry
> decoratedSupplier = Retry
>     .decorateSupplier(retry, decoratedSupplier);
> 
> // Use of Vavr's Try to
> // execute the decorated supplier and recover from any exception
> String result = Try.ofSupplier(decoratedSupplier)
>     .recover(throwable -> "Hello from Recovery").get();

Has anyone used CheckedSupplier successfully in this scenario?

**Comment by ghost** [comment]
Yes, worked for me. Just use CheckedSupplier instead of CheckedFunction0 and use .get() instead of .apply() method since it is a Supplier and Not a Function anymore.
Hope this helps. Thanks!!
